kotlinx-coroutines-core/common/src/sync/Mutex.kt

```diff
@@ -355,24 +349,28 @@ internal class MutexImpl(locked: Boolean) : Mutex, SelectClause2<Any?, Mutex> {
     }
 
     private class LockCont(
+        mutex: MutexImpl,
         owner: Any?,
         @JvmField val cont: CancellableContinuation<Unit>
-    ) : LockWaiter(owner) {
-        override fun tryResumeLockWaiter() = cont.tryResume(Unit)
+    ) : LockWaiter(mutex, owner) {
+        override fun tryResumeLockWaiter() = cont.tryResume(Unit, idempotent = null) {
+            // if this continuation get's cancelled during dispatch to the caller, then release the lock
+            mutex.unlock(owner)
+        }
```

Comment:
We allocate an extra lambda on each resume of the lock waiter. This might be Ok (this is kind of slow path) but needs additional thought.

Reply:
Does not really make a big difference.